### ISSUE TITLE
Subtyping rules with functions with dependently-typed signatures

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-spec#633]
**Description:**
Given the following class definitions

```ballerina
class Foo {
    function get(typedesc<anydata> td) returns td|error = external;
}

class Bar {
    function get(typedesc<anydata> td) returns anydata|error => "hello world";
}
```

I assume `Foo` can be considered to be a subtype of `Bar` here.

But can `Bar` be considered a subtype of `Foo`? While the value space containing all the values that can be returned by the two methods are the same, considering `Bar` to be a subtype of `Foo` can lead to issues at runtime?

```ballerina
Foo foo = new Bar();

int|error result = foo.get(int); // actually returns a `string` value at runtime
```

**Comment by jclark** [comment]
I agree `Foo` is a subtype of `Bar`, but they other way round is not the case.

You can't do function subtyping just in terms of sets of values because of #86, and we can only really do function subtyping properly when we deal with that, which also really needs to be done in conjunction with implementing structural typing properly.

For now, I suggest we recognize that a dependently-typed return type is a subtype of the corresponding non-dependently typed value (i.e. recognize `Foo` as a subtype of `Bar`).

Can you suggest some other easy cases we can recognize that are sound?

**Comment by MaryamZi** [comment]
I guess the case where the typedesc param allows a singleton will work, although probably not quite useful.

```ballerina
const BAZ = "baz";

class Foo {
    function get(typedesc<BAZ> td) returns td = external;
}

class Bar {
    function get(typedesc<BAZ> td) returns BAZ => BAZ;
}
```

Even with distinct types, IINM we can run into the original issue with scenarios where type-ids are induced.

**Comment by MaryamZi** [comment]
I think there's also a related issue with scenarios where both the functions/methods have dependently-type signatures and the typedesc params are of different types.

**Case I: typedesc params of the same type**
I assume if the typedesc params are of the same type, and the rest of the normal function subtyping constraints are satisfied, one can be considered a subtype of the other.

e.g., in the following scenario `Bar` can be considered a subtype of `Foo`.

```ballerina
class Foo {
    function get(typedesc<anydata> td) returns td|error = external;
}

class Bar {
    function get(typedesc<anydata> td) returns td = external;
}
```

**Case II: typedesc params of different types**
What if the relevant typedesc param in one of the types is of a broader type?

For example,

```ballerina
class Foo {
    function get(typedesc<anydata> td) returns td = external;
}

class Bar {
    function get(typedesc<any> td) returns td = external;
}
```

Here, `Foo` is not a subtype of `Bar`.

But given that `Bar`'s `get` method is depenently-typed, guarantees that it will return a relevant `anydata` value if the argument passed for `td` is a subtype of `typedesc<anydata>` in-line with the contract defined by `Foo`, should we consider `Bar` as a subtype of `Foo`?

**Comment by jclark** [comment]
Yes for Case I.

No for Case II, because argument subtyping is contravariant i.e. Bar's a subtype of Foo as regards the return type but not as regards the parameter types.

**Comment by MaryamZi** [comment]
Since subtyping for parameters is based on contravariance, isn't `Bar` a subtype of `Foo` (in terms of the parameter) since `typedesc<anydata>` is a subtype of `typedesc<any>`? Is it not the return type that causes the conflict?

Despite the possible difference in return values, IINM, allowing `Bar` as a subtype of `Foo` would not cause any surprises  at runtime, and will return values of the expected types.

```ballerina
public function main() {
    Foo f = new Bar();

    // `f.get()` will always behave as expected even though it calls `get` of `Bar`.
    anydata v1 = f.get(anydata); 
    int v2 = f.get(int);

    // Since we can never pass a value that does not belong to `typedesc<anydata>` as the 
    // argument, the returned value will also always belong to `anydata`.
    any v3 = f.get(any); // compilation error: incompatible types: expected 'typedesc<anydata>', found 'typedesc<any>'
}
```

**Comment by jclark** [comment]
Sorry, I thought you meant the other way round.

I think you're right that treating Bar as a subtype of Foo would be sound, i.e. if you do things that are allowed on a Foo on a Bar, it will behave like a Foo.

However, I have no idea how I would explain this in terms of structural typing, and given the restriction that dependently-typed functions have to be external, I would propose postponing this until we do #86.

**Comment by MaryamZi** [comment]
Noted. For now, we will allow just the straightforward scenarios in the implementation.

**Comment by MaryamZi** [comment]
I believe the rules would also have to cover the scenario where there are multiple typedesc params.

e.g.,
```ballerina
class Foo {
    function get(typedesc<anydata> x, typedesc<anydata> y) returns x|error = external;
}

class Bar {
    function get(typedesc<anydata> y, typedesc<anydata> x) returns x|error = external;
}

class Baz {
    function get(typedesc<anydata> y, typedesc<anydata> x) returns y|error = external;
}
```

Given that the parameter name is irrelevant in function typing, I believe we could look at the the index of the parameter (referred in the return type) in the list of parameters.

So in this scenario `Foo` will be a subtype of `Baz` and vice-versa, but `Bar` will not be a subtype of `Foo` (or `Baz`) nor will `Foo` (or `Baz`) be a subtype of `Bar`.

**Comment by jclark** [comment]
I agree.
